drivers/i2s/i2s_cavs.c

```diff
@@ -602,79 +615,108 @@ static int i2s_cavs_trigger(struct device *dev, enum i2s_dir dir,
 	unsigned int key;
 	int ret;
 
-	if (dir == I2S_DIR_TX) {
-		strm = &dev_data->tx;
-	} else {
-		SYS_LOG_ERR("TX direction must be selected");
-		return -EINVAL;
-	}
+	strm = (dir == I2S_DIR_TX) ? &dev_data->tx : &dev_data->rx;
 
+	key = irq_lock();
 	switch (cmd) {
 	case I2S_TRIGGER_START:
 		if (strm->state != I2S_STATE_READY) {
-			SYS_LOG_DBG("START trigger: invalid state");
+			irq_unlock(key);
+			SYS_LOG_ERR("START trigger: invalid state %u",
+					strm->state);
 			return -EIO;
 		}
 
 		__ASSERT_NO_MSG(strm->mem_block == NULL);
 
-		ret = strm->stream_start(strm, ssp, dev_data->dev_dma);
+		if (dir == I2S_DIR_TX) {
+			ret = i2s_tx_stream_start(dev_data, ssp,
+					dev_data->dev_dma);
+		} else {
+			ret = i2s_rx_stream_start(dev_data, ssp,
+					dev_data->dev_dma);
+		}
+
 		if (ret < 0) {
+			irq_unlock(key);
 			SYS_LOG_DBG("START trigger failed %d", ret);
 			return ret;
 		}
 
 		strm->state = I2S_STATE_RUNNING;
-		strm->last_block = false;
 		break;
 
 	case I2S_TRIGGER_STOP:
-		key = irq_lock();
 		if (strm->state != I2S_STATE_RUNNING) {
 			irq_unlock(key);
 			SYS_LOG_DBG("STOP trigger: invalid state");
 			return -EIO;
 		}
 		strm->state = I2S_STATE_STOPPING;
-		irq_unlock(key);
-		strm->last_block = true;
 		break;
 
 	case I2S_TRIGGER_DRAIN:
-		key = irq_lock();
 		if (strm->state != I2S_STATE_RUNNING) {
 			irq_unlock(key);
 			SYS_LOG_DBG("DRAIN trigger: invalid state");
 			return -EIO;
 		}
 		strm->state = I2S_STATE_STOPPING;
-		irq_unlock(key);
 		break;
 
 	case I2S_TRIGGER_DROP:
 		if (strm->state == I2S_STATE_NOT_READY) {
+			irq_unlock(key);
 			SYS_LOG_DBG("DROP trigger: invalid state");
 			return -EIO;
 		}
-		strm->stream_disable(strm, ssp, dev_data->dev_dma);
-		strm->queue_drop(strm);
+		if (dir == I2S_DIR_TX) {
+			i2s_tx_stream_disable(dev_data, ssp, dev_data->dev_dma);
+		} else {
+			i2s_rx_stream_disable(dev_data, ssp, dev_data->dev_dma);
+		}
 		strm->state = I2S_STATE_READY;
 		break;
 
 	case I2S_TRIGGER_PREPARE:
 		if (strm->state != I2S_STATE_ERROR) {
+			irq_unlock(key);
 			SYS_LOG_DBG("PREPARE trigger: invalid state");
 			return -EIO;
 		}
 		strm->state = I2S_STATE_READY;
-		strm->queue_drop(strm);
 		break;
 
 	default:
+		irq_unlock(key);
 		SYS_LOG_ERR("Unsupported trigger command");
 		return -EINVAL;
 	}
 
+	irq_unlock(key);
+	return 0;
+}
+
+static int i2s_cavs_read(struct device *dev, void **mem_block, size_t *size)
+{
+	struct i2s_cavs_dev_data *const dev_data = DEV_DATA(dev);
+	struct stream *strm = &dev_data->rx;
+	void *buffer;
+	int ret = 0;
+
+	if ((strm->state == I2S_STATE_NOT_READY) ||
+		(strm->state == I2S_STATE_ERROR)) {
+		SYS_LOG_ERR("invalid state %d", strm->state);
+		return -EIO;
+	}
+
+	ret = k_msgq_get(&strm->out_queue, &buffer, dev_data->cfg.timeout);
```

Comment:
should this be in_queue? Looks like that if one looks at the rx_callback function. If that's true, i am not really sure why out_queue is needed (for both playback and recording).

Reply:
> should this be in_queue? Looks like that if one looks at the rx_callback function. If that's true, i am not really sure why out_queue is needed (for both playback and recording).

I've added some comments above the definition of `struct stream`. Please check, I hope that it clarifies the usage of in and out queues.

Also see #10948 which when address should help in reducing the number of queues.